graylog2-web-interface/src/views/components/visualizations/ChartData.js

```diff
@@ -10,64 +10,81 @@ export type ChartDefinition = {
   name: string,
   x?: Array<string>,
   y?: Array<any>,
+  z?: Array<Array<any>>,
 };
 
-export type ChartData = [any, Array<Key>, Array<any>]
+export type ChartData = [any, Array<Key>, Array<any>, Array<Array<any>>]
 export type ExtractedSeries = Array<ChartData>;
 
 export type KeyJoiner = (Array<any>) => string;
 
-export type Generator = (string, string, Array<string>, Array<any>, number, number) => ChartDefinition;
+export type Generator = (string, string, Array<string>, Array<any>, Array<Array<any>>, number, number) => ChartDefinition;
 
 const _defaultKeyJoiner = keys => keys.join('-');
 
-export const extractSeries = (keyJoiner: KeyJoiner = _defaultKeyJoiner) => {
-  return (results: Rows): ExtractedSeries => {
-    // $FlowFixMe: Somehow flow is unable to infer that the result consists only of Leafs.
-    const leafs: Array<Leaf> = results.filter(row => (row.source === 'leaf'));
+const _defaultChartGenerator = (type, name, labels, values): ChartDefinition => ({ type, name, x: labels, y: values });
 
-    const x = leafs.map(({ key }) => key);
+const _flattenLeafs = (leafs: Array<Leaf>, showSummarySeries: boolean = true) => {
```

Comment:
When we currently create a chart with the metric count, e.g. for a linechart, we always display a "line" for a summary of all counts. This would make the heatmap useless. This is the best approach that came through. Any ideas of a different approach? Or a different code style?

Reply:
I would replace it with a matcher instead. The default would look like `value => value.source.endsWith('leaf')` while you would override it in your use case with `value => (value.source === 'row-leaf')`. This makes it way more flexible and in the function's scope the abstraction level is not left (extracting leafs, vs. summary series).

Reply:
Thanks, I really like this approach. I've implemented the matcher function and added a test.